### ISSUE TITLE
OpenAPI Workspace Feature

### ISSUE THREAD

**Issue by darrelmiller** [microsoft/OpenAPI.NET#477]
OpenAPI allows reusing existing definitions using the $ref syntax.  The current implementation only supports what I am calling "local" references.  This means the $ref points to a component in the current OpenAPI document.

The OpenAPI specification also supports referencing content that exists in external documents.  There are two types of Remote references that use the $ref syntax and a third that uses url typed properties:
- A reference to a component in an external document that is a valid OpenAPI document.
- A reference to an external document that contains a fragment of JSON or YAML that should be interpreted based on the context of the reference. (fragment)
- A reference to an external document that contains come stream of bytes. This could be an example value, a license document, a terms of service document, or some kind of contact information. (artifact)

## Work done so far:

- Introduced a new class OpenApiWorkspace that can contain multiple instances of OpenAPIDocuments, fragments and artifacts.
- Added a ResolveReference method to OpenAPIDocument to enable resolving local references.
- Created a ResolveReference method on OpenAPIWorkspace to resolve remote references or defer to OpenApiDocument to resolve local references.
- Added a BaseUrl to OpenAPIWorkspace to support relative remote references.  Use file:// url to point to a file system.

Current work is ongoing in branch: dm/OpenAPIWorkspace

## Work remaining:
- Load document fragments and reference them
- Load artifacts and reference them

## Unknowns:
- Do we need the notion of a "primary document"?

## Future work
- Export all items in a OpenAPIWorkspace as zip or multipart
- Merge OpenApiDocuments and fragments into a single document without remote references
- Inline all references (except artifacts)

**Comment by simon-pearson** [comment]
Hi Darrel, I'd like to contribute towards this feature if possible.  For now I'll focus my efforts on adding support for references to remote document fragments.

Can we clarify how the client code will use document fragments?  I've added the following failing tests to `OpenApiWorkspaceTests.cs` which I think cover the main ways in which a client would want to reference a document fragment:
```
[Fact]
public void OpenApiWorkspacesCanResolveReferencesToDocumentFragments()
{
    // Arrange
    var workspace = new OpenApiWorkspace();
    var schemaFragment = new OpenApiSchema { Type = "string", Description = "Schema from a fragment" };
    workspace.AddFragment("fragment.json", schemaFragment);

    // Act
    var schema = workspace.ResolveReference(new OpenApiReference()
    {
        Id = "test",
        Type = ReferenceType.Schema,
        ExternalResource = "fragment.json"
    }) as OpenApiSchema;

    // Assert
    Assert.NotNull(schema);
    Assert.Equal("From a fragment", schema.Description);
}

[Fact]
public void OpenApiWorkspacesCanResolveReferencesToDocumentFragmentsWithJsonPointers()
{
    // Arrange
    var workspace = new OpenApiWorkspace();
    var componentsFragment = new OpenApiComponents
    {
        Schemas =
        {
            ["schema1"] = new OpenApiSchema() {Type = "string", Description = "Schema from a fragment"}
        }
    };
    workspace.AddFragment("fragment.json", componentsFragment);

    // Act
    var schema = workspace.ResolveReference(new OpenApiReference()
    {
        Id = "test",
        Type = ReferenceType.Schema,
        ExternalResource = "fragment.json#/components/schemas/schema1"
    }) as OpenApiSchema;

    // Assert
    Assert.NotNull(schema);
    Assert.Equal("Schema from a fragment", schema.Description);
}
```
**Questions**
1\. What about document fragments which are valid JSON, and composed of valid OpenAPI objects, but are not themselves a valid OpenAPI object.  For example consider the following JSON document fragment:
```
{
	"SchemaOne": {
		"type": "object",
		"required": [
			"id"
		],
		"properties": {
			"id": {
				"type": "integer",
				"format": "int64"
			}
		}
	},
	"SchemaTwo": {
		"type": "object",
		"required": [
			"name"
		],
		"properties": {
			"name": {
				"type": "string"
			}
		}
	}
}
```
`SchemaOne` and `SchemaTwo` are both valid Schema objects themselves, but the top-level JSON object with the `SchemaOne` and `SchemaTwo` properties is not itself a valid OpenAPI object.  Currently the `OpenApiWorkspace.AddFragment` method has the following signature:

```void AddFragment(string location, IOpenApiElement fragment)```

By declaring the `fragment` argument as an `IOpenApiElement` we cannot support such 'free-form' document fragments.  Do you have any thoughts around this?

2\.  Does the above capture how you envisaged clients adding document fragments to workspaces? 
 Are there any usages missing from the above?

3\.  Did you have any thoughts about how to parse the JSON reference to retrieve the particular part of a document fragment file?

**Comment by darrelmiller** [comment]
Question1:   This is a great question and one I alluded to in a comment in the source:
```
  /// <remarks>Not sure how this is going to work.  Does the reference just point to the fragment as a whole, or do we need to 
  /// to be able to point into the fragment.  Keeping it private until we figure it out.
  /// </remarks>
```
I think we need to allow pointing to a valid object embedded somewhere within an arbitrary JSON/YAML fragment.

Question 2:
The tests make sense to me.  The second one is a bit ambiguous in that you can't tell fragment.json is actually a valid OpenAPI description or not.

Question 3:
No. I'm completely open to suggestions.

**Comment by simon-pearson** [comment]
Darrel,

Many thanks for merging PR #520.
I'm just starting to take a look at referencing artifacts now.  Questions:
1. In the above description you describe artifacts as "A reference to an external document that contains some stream of bytes that can be used as an example".  If artifacts are only to be used as examples, is it fair to say that all artifacts will ultimately be classes which implement `IOpenApiAny`?
1. Is it valid to reference a nested property of an `IOpenApiAny` using a JSON pointer? e.g. if the artifact is an instance of `OpenApiObject` which has a property which is itself an `OpenApiObject`, is it possible to reference through to the nested property via a JSON pointer?
1. When resolving a reference to an artifact I assume I'm going to have to deserialize the stream back to an object implementing `IOpenApiAny`?  If so, how will I know which deserializer to use (JSON/YAML)?  One option would be to add an `OpenApiFormat` argument to the `OpenApiWorkspace.AddArtifact(...)` method to record the format of each artifact added to the workspace.  What are your thoughts?

**Comment by darrelmiller** [comment]
Hey @simon-pearson,

It's starting to get redundant to apologize for taking so long to reply, so let's just assume that by default.

I have finally finished the merge of vnext into dm/openapiworkspace and I'm currently debating if I just just bite the bullet and merge it into vnext.  I also, have news regarding my previously constrained view of external references.  We had a conversation at the TSC meeting this week and we are going to add some clarifying words to the spec.  But first your questions.

1) Artifacts don't need to be JSON or YAML.  In fact that's the primary advantage of being able to point to examples in an external file.  Artifacts used for examples could be XML, CSV, JPG, PDF, etc.  I would say artifacts should be represented as a stream.  That way we can hold references to a memory stream or a file stream.  I wasn't clear in the description of this issue (possibly because I had forgot) but artifacts were intended to be used for content that are not referred to by $ref.  It's for things like example.externalValue, license.url, info.termsOfService, contact.url.

2) Ugh.  The spec isn't precise.  For externalValue in example it just says "A URL that points to a Iiteral example".  I could go either way here.  I'd be happy to just say no initially and consider adding it if people ask for it.  

3)  I think by now you have realized that I confused the issue as to what an artifact is.  I think remote documents and fragments handle all cases where external references point to JSON or YAML.  Artifact is specifically for non-YAML/JSON content.

**Comment by darrelmiller** [comment]
Here's the proposed new wording around the use of $ref
> A simple object to allow referencing other objects in the OpenAPI document, internally and externally. Targets of a reference do not need to be contained in a components section and for external references, targets MAY exist within any compatible resource. Referenced objects are subject to the same constraints as inline objects.

https://github.com/OAI/OpenAPI-Specification/pull/2435/files

This means that $ref can point to objects that are anywhere in a document.  So, allow me to apologize for having you remove code that previously supported that.  I'm not a fan of allowing this, but I was outvoted  :-)

**Comment by simon-pearson** [comment]
Darrel, a couple of questions for you, one on fragments and one on artifacts:

### Question 1. Document Fragments
> Here's the proposed new wording around the use of $ref
> 
> > A simple object to allow referencing other objects in the OpenAPI document, internally and externally. Targets of a reference do not need to be contained in a components section and for external references, targets MAY exist within any compatible resource. Referenced objects are subject to the same constraints as inline objects.
> 
> https://github.com/OAI/OpenAPI-Specification/pull/2435/files
> 
> This means that $ref can point to objects that are anywhere in a document. So, allow me to apologize for having you remove code that previously supported that. I'm not a fan of allowing this, but I was outvoted :-)

No problem, "change is the only constant" 😀.  I'm not sure I fully understand the implications of this, though.  Are you saying that it is valid to have, for example, a fragment containing a top-level [Operation object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#operationObject) (which is not in itself referencable), but then reference it's child properties?

e.g.

```
[Fact]
public void OpenApiWorkspacesCanResolveReferencesToNonTopLevelElementsInADocumentFragment()
{
    // Arrange
    var workspace = new OpenApiWorkspace();
    var operationFragment = new OpenApiOperation()
    {
        Parameters = new List<OpenApiParameter>
        {
            { new OpenApiParameter() }
        }
    };
    workspace.AddFragment("fragment", operationFragment);

    // Act
    var resolvedElement = workspace.ResolveReference(new OpenApiReference()
    {
        Id = "parameters/0",
        ExternalResource = "fragment"
    });

    // Assert
    Assert.Same(operationFragment.Parameters[0], resolvedElement);
}
```

If the above is valid then this will mean we'll want to move our `ResolveReference` extension method on to `IOpenApiElement` rather than `IOpenApiReferenceable`, right?  If you confirm the above understanding is right then I'll set about making the relevant changes.

### Question 2. Artifacts
> I think by now you have realized that I confused the issue as to what an artifact is. I think remote documents and fragments handle all cases where external references point to JSON or YAML. Artifact is specifically for non-YAML/JSON content.

I've added the following unit test which exercises adding an artifact to a workspace and retrieving it:

```public void OpenApiWorkspacesCanStoreAndReferenceDocumentArtifacts()
{
    // Arrange
    var stream = new MemoryStream(ASCIIEncoding.Default.GetBytes("Some stream of bytes"));
    var workspace = new OpenApiWorkspace();
    workspace.AddArtifact("artifact", stream);

    // Act
    var workspaceArtifact = workspace.GetArtifact("artifact");

    // Assert
    Assert.NotNull(workspaceArtifact);
    Assert.Same(stream, workspaceArtifact);
}
```
I'm struggling to grasp how clients will use artifacts.  Is it possible to reference a document artifact using a `OpenApiReference`?  If we had a [Media Type Object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#media-type-object) and we wanted to reference a `Stream` artifact for the `example` property, how would it do this?  Or am I still a bit off point here and not quite capturing the intended use of artifacts.
TL;DR: can you provide a brief snippet of code showcasing how you envisage artifacts being used?

**Comment by darrelmiller** [comment]
1. Document Fragments
The solution you are proposing would allow us to parse document fragments that are not complete OpenAPI documents, but contain some reusable OpenAPI element within them.  However, it goes further.  The valid reusable OpenAPI component can exist in a JSON/YAML document that itself is not a valid OpenAPI description.  e.g.  Someone could create a file called sharedparameters.yaml that looks like this:

```yaml
- name: foo
  in: query
  schema:
    type: string
- name: bar
   in: query
   schema:
     type: string
```

I suspect we need to introduce a new type `OpenApiFragment` that can be used to extract OpenApiReferencable items.  Internally I don't know if we just hold on to it as a YamlStream, or a collection of ParseNodes.  It's kind of annoying to support this because it means that we can't fully process it into a DOM.  We have to leave it in a partially processed state.

Question 2: Artifacts

I think we might add helper methods to OpenApiExample and OpenApiExternalDocs to GetArtifact().  This library really isn't going to do much with the artifact.  The primary role is to resolve the URL and ensure when we package up the Workspace into a zip or multi-part content, that the artifacts get included and have the right filenames.

Anyway, that's the only function I have considered.  Other tooling like Validators or mocking servers would use the artifacts for more things.

**Comment by darrelmiller** [comment]
For the moment, I'm not going to touch the fragment issue.  I want to spend some time working on enabling Async reading.  We can't deal with resolving external documents without adding ReadAsync support.

**Comment by darrelmiller** [comment]
1.3 is now successfully using workspaces for external references to other OpenAPI files.  The complete implementation will come in the v2 version as breaking changes will be required for a complete implementation of external references.
